README.md
VEXTREME Vectrex Multicart
===

We're carrying on and pushing forward the amazing work of [Sprite_tm](http://spritesmods.com/?art=veccart&page=1) and his Extreme Vectrex Multicart. He posted the code (GPLv3 License) and lots of screens of the PCB/initial schematic, but he never released the PCB. It's a great base for an inexpensive and open vectrex multicart, for developers and players alike!

:warning: Please read this NOTICE of Current Development :warning:
===

**This project is a work in progress, and is not ready for mass production.**  Please observe the currently [posted issues](https://github.com/technobly/vextreme/issues) and there may be more lurking that are not yet identified.  The hardware and software is evolving.  You are free to build any version, contribute and/or follow along.  Please note that support of these pre-releases will be very limited and if you embark on an epic adventure to build one or a few, you should feel comfortable with this type of embedded hardware and software development.  There will be road bumps, but the journey is half the fun of the destination.  When things appear to be stable and we've worked out most of the kinks, we'll change the version to v1.0.  Only then is it advisable to mass produce.

3D Renders
===

![v0.2 Render](images/vextreme-v0.2-white.gif)

Video Updates
===

Be sure to subscribe to my Youtube channel for updates!  Here's my [Vectrex playlist](https://www.youtube.com/watch?v=zkJ-z77fJCw&list=PL7MgXfpGKg6CCaIQejZVb5CsTg3-5mHU0&index=1) if you want to stay laser focused on this project.

VEXTREME Discord Server
===
Join the [VEXTREME Discord server](https://discord.gg/VDssGVJ) to chat with us about the development

BOM and Parts ordering
===

All the parts are described in Bill of Materials [vextreme-v0.x.csv](bom/)

Also, you can use this [Digi-Key shared cart](https://www.digikey.com/short/zpdmtp) if you're in a hurry.  It has will have part references right on the packages for you!

Ordering PCB's
===

OSHPark is a good place to order with purple or the new "after dark" theme color scheme.  You can upload the KiCad [vextreme.kicad_pcb](hardware/vextreme.kicad_pcb) there directly.  I would download this entire Github repo ZIP file first though instead of just trying to save the PCB file from your browser.

Another way to order PCB's is by using the included [gerbers-vextreme-v0.x.zip](gerbers) and uploading those with all of the necessary specs to companies like [PCBWay](https://www.pcbway.com) or [JLCPCB](https://jlcpcb.com)

Building and Flashing STM firmware
===

### This assumes you have these prerequisits installed:

- Docker
- dfu-util v0.9
- Make
- Windows will also require these [Zadig instructions](https://github.com/profezzorn/ProffieOS/wiki/zadig)

### Build the stm32-build docker image
```
code/stm32 $ make docker-build
```

### Connect the 2-pin jumper and then connect the cart to USB

### Check if your computer sees the STM32
```
$ dfu-util -l

dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2019 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Deducing device DFU version from functional descriptor length
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=3, name="@Device Feature/0xFFFF0000/01*004 e", serial="123412341234"
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=2, name="@OTP Memory /0x1FFF7800/01*512 e,01*016 e", serial="123412341234"
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=1, name="@Option Bytes  /0x1FFFC000/01*016 e", serial="123412341234"
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=0, name="@Internal Flash  /0x08000000/04*016Kg,01*064Kg,03*128Kg", serial="123412341234"
```

### Build and flash the STM32 image via dfu-util
```
code/stm32 $ make clean all flash USE_HW=v0.3

// NOTE: Use v0.2 above if you have v0.2 HW, or you may use v0.3 if you make the proper mods to your board.  See "Modifying v0.2 to v0.3"

// You should end up with something like this

Downloading to address = 0x08000000, size = 23536
Download    [=========================] 100%        23536 bytes
Download done.
File downloaded successfully
```

### Remove the cart from USB and remove the jumper (install it on one pin so you don't lose it)


Building and Flashing the multicart's menu
===

### Build the asm6809 docker image
```
code/menu $ make docker-build
```

### Connect the cart to USB with the 2-pin jumper removed

- The first time you do this you will likely have to format the USB drive

  - Name: VEXTREME
  - Format: MS-DOS (FAT)
  - Scheme: Master Boot Record
  - 512 byte page size

- Be patient, it's slow!  Don't worry, copying binaries later will be fast!!

### Build and flash the multicart menu binary
```
// Mac OS
code/menu $ make clean all && cp menu.bin /Volumes/VEXTREME/

// optionally add this to the above command to unmount, after you figure out which drive it is
$ diskutil list
&& diskutil unmountDisk /dev/disk3

// Linux
code/menu $ make copy

// You should end up with something like this, and the binary should have been copied to the root of your multicart drive

asm6809  -B -o menu.bin menu.asm
```

Modifying v0.2 to v0.3
===

![v0.3 Mod 1](images/v0.3-mod1.png)

- To get the new reset feature of v0.3, as of v0.24 software you will need to jumper V-OE pin 12 of the cart fingers (or U3 pin 9) to STM32 pin 29 (PB10).  This may change in the future, don't sell your soldering iron!
- The wire mod above also allows you to play Animaction on VEXTREME!
- More TBD


Adding Vectrex Game ROMs (Binaries)
===

- Create a directory on your multicart called `roms/`

- Add your first binaries there, may I suggest: [Beluga Dreams](https://8-bit-waves.itch.io/beluga-dreams) and [Vec-Man](http://eiti.fh-pforzheim.de/personen/johannsen/projektlabor/vectrex_2019/vectrex_game_vec_man.htm) ?

Docs
===

A nice collection of lessons learned will be [documented here.](docs/)


Hall of Fame
===

If you build one of these, be sure to send me a message with your build pics showing your working VEXTREME multicart!  Or better yet, check out the [hall of fame format here](hall-of-fame/) and submit a PR (please and thank you!)

LICENSE
===

GPLv3 - essentially you must keep this open, all changes must be disclosed and shared.  Full [LICENSE here](LICENSE) and [TL;DR summary here](https://tldrlegal.com/license/gnu-general-public-license-v3-(gpl-3))


CHANGELOG.md
# Changelog

## v0.24

Hardware Changes

- rename 'veccart' to 'vextreme' for hardware
- add image for v0.3 wire mod
- Fixes #17, make all pads on addressable LEDS the same size
- Fixes #10 removes soldermask ink around PCB fingers
- Fixes #11 Edge of PCB fingers should be 18.0mm from center of 3.52mm mounting hole and 1.2mm from edge, and slightly enlarged (0.2mm oversized)
- Also aligns holes, fingers and board outline on 0.1mm grid

Software Changes

- Added Turkey Pop.bin to release
- Added HACKS/LEVEL SELECT/Mine Storm 1.bin to release
- Fixed GCE/Mine Storm 1.bin in release
- rename 'veccart' to 'stm32' for firmware, and rename 'multicart' to 'menu'
- Embedded menu binary in STM32 flash image
- Update Dockerfile distro to Arch w/ newer arm-gcc (10.1)
- add .map file to build output
- Dev Mode (cart.bin) and Menu (multicart.bin) should not load a high score
- made high_score_flag readable from parmRam with new RPC ID 16, and speeds memory r/w for highscore
- Implements #55 High Score (by @TylerBrymer PR #56)
- Truncate filenames displayed to 16 characters and fix long filename bug
- Reduced rainbow mode colors so that no-buzz Vectrex owners don't hear any LED modulation
- implemented Animaction support for v0.3 hardware
- made sys_opt.hw_ver usage clearer
- added RAM Write debugging application and new dumpMemory() RPC ID 13
- added a release.sh script
- added standalone Developer Mode app (note: this is for reference only, see readme.md)
- added Develop Mode docs
- added Developer Mode (DM) to Menu.  Press button 2 & 3 at any time to enter DM, or hold them down when starting the Menu.  DM will also automatic re-enter when it's active.  DM is active when a cart.bin is present, and DM can be disabled by removing cart.bin manually, or through the DM menu button 1: Exit.
- add USE_HW compile option, which must now be set.  It will not default to a specific version.  Right now specify it as USE_HW=v0.2 or USE_HW=v0.3 when running 'make clean all flash'.  It can also be exported in your environment if you wish, but don't forget about it if you update your HW ;-)
- removed :leave from DFU commands since after flashing the stm32 binary, we are not really ready to reset and run.  The boot0 jumper must be removed before then.
- minor changes to improve the Makefile for Windows (still WIP)
- changed delay() and millis() to use dwt_read_cycle_counter() instead of the sys_tick_handler() because romemu.S didn't like us randomly leaving to handle interrupts, and we need these functions while we are emulating ROM.  There is a ~35s delay limit because of this.  I added an assert so I don't forget!
- started adding some changes in fatfs to enable writes to flash memory since it was currently read-only, but ended up not needing them.  No worries though, the highscore feature will require these.  These changes thanks to @tbrymer!
- VEXTREME now also prints it's own HW and SW revision in logs at boot time.
- Moved start of rainbow LED code to after we are done with USB activities, so as not to load down the USB port more than necessary.  Hopefully this fixes a reported bug with Windows.
- USB MSC can now be entered while the Vectrex is off, Ejected and left plugged in and the Vectrex can be turned on and boot a cart.bin for Developer Mode, or the VEXTREME menu.
- USB MSC can be started and stopped from the Vectrex now, which is necessary for Developer Mode.
- An improved reset function was added after feedback from users.  Now in v0.3 hardware you will be able to press reset at any time to get back to the Menu.  If you hold reset for ~700ms, it will reset the current running ROM/game.  In the case of booting into Developer Mode (cart.bin already on the drive root), a first reset will not go to the Menu, but rather the cart.bin.. allowing you to skip the cold boot sequence. You may modify v0.2 hardware to act like v0.3 hardware for this feature.  For now that means you will need to jumper V-OE pin 12 of the cart fingers (or U3 pin 9) to STM32 pin 29 (PB10).
- For v0.2 hardware, you can still compile in the old functionality were a short reset would reset the running ROM/game, and if you hold reset for ~700ms you will get the Menu.
- Also added to the reset sequence for both v0.2/v0.3 HW is that the LEDs light up CYAN after the "long" reset delay of 700ms, and stay lit as long as you are holding reset.  When you let go, they return to rainbow.  This helps you gage just how long you need to wait to hold the button for long reset.
- Added "scsi_command: %02X\n" which logs USB MSC commands to log output, this might seem chatty, but for now let's see what we see ;-)
- Set USB power to 500mA (was set to 100mA)

## v0.23

- Fixes issue where VEXTREME takes several minutes to mount as USB drive on Windows #53

## v0.22

- ignore all file extensions that are not .bin or .vec (case insensitive) #49
- increased quick reset delay to 700ms #45 
- adds HW/SW version to VEXTREME menu and games #48 
  - see example in LED Test v1.1.bin

## v0.21

- :warning: Added NOTICE of Current Development
- Fixes Polar Rescue and implements a new Reset Heurisic #37
- Adds addressable LED control and demo #41
    - added 3 functions to control the LEDs from a Vectrex program
        - updateOne() - RPC 6
        - updateAll() - RPC 4
        - updateMulti() - RPC 7
    - added RPC function 5 to step through rainbow control of LEDs in menu
    - added "LED TEST" which uses updateMulti to show how to finely control
    all of the LEDs with a test program.  Press UP on the joystick, try it slowly as well :D
    - added addressable LED control to Veccy Bird v1.5

## v0.2

- Replaced USB-B-mini with USB-C and centered USB on cart
- Extended the PCB by 1.7 mm to get the USB-C connector as close to the case exterior as possible.
- Widened the PCB to 48.0 mm so that there’s less side to side play in the cartridge slot.
- Moved outer mounting holes into the proper locations

code/stm32/led.h
/*------------------------------------------------------------------------
  Ported to STM32F411 for VEXTREME by Technobly

  ------------------------------------------------------------------------
  -- original header follows ---------------------------------------------
  ------------------------------------------------------------------------

  This file is part of the Adafruit Dot Star library.

  Adafruit Dot Star is free software: you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public License
  as published by the Free Software Foundation, either version 3 of
  the License, or (at your option) any later version.

  Adafruit Dot Star is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with NeoPixel.  If not, see <http://www.gnu.org/licenses/>.
  ------------------------------------------------------------------------*/

#ifndef _ADAFRUIT_DOT_STAR_H_
#define _ADAFRUIT_DOT_STAR_H_

// Color-order flag for LED pixels (optional extra parameter to constructor):
// Bits 0,1 = R index (0-2), bits 2,3 = G index, bits 4,5 = B index
#define RGB_RGB (0 | (1 << 2) | (2 << 4))
#define RGB_RBG (0 | (2 << 2) | (1 << 4))
#define RGB_GRB (1 | (0 << 2) | (2 << 4))
#define RGB_GBR (2 | (0 << 2) | (1 << 4))
#define RGB_BRG (1 | (2 << 2) | (0 << 4))
#define RGB_BGR (2 | (1 << 2) | (0 << 4))

// Addressable LED Preset Colors
static const uint32_t colors[10] = {
    0,        // off
    0xFF0000, // red
    0xFF9900, // orange
    0xFFFF00, // yellow
    0x00FF00, // green
    0x00FFFF, // cyan
    0x0000FF, // blue
    0x7700FF, // pink
    0xFF00FF, // magenta
    0xFFFFFF  // white
};

/* 8-bit gamma-correction table.
   Copy & paste this snippet into a Python REPL to regenerate:
import math
gamma=2.6
for x in range(256):
    print("{:3},".format(int(math.pow((x)/255.0,gamma)*255.0+0.5))),
    if x&15 == 15: print
*/
static const uint8_t _LedsGammaTable[256] = {
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  1,  1,  1,
    1,  1,  1,  1,  2,  2,  2,  2,  2,  2,  2,  2,  3,  3,  3,  3,
    3,  3,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,  7,
    7,  7,  8,  8,  8,  9,  9,  9, 10, 10, 10, 11, 11, 11, 12, 12,
   13, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 20,
   20, 21, 21, 22, 22, 23, 24, 24, 25, 25, 26, 27, 27, 28, 29, 29,
   30, 31, 31, 32, 33, 34, 34, 35, 36, 37, 38, 38, 39, 40, 41, 42,
   42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
   58, 59, 60, 61, 62, 63, 64, 65, 66, 68, 69, 70, 71, 72, 73, 75,
   76, 77, 78, 80, 81, 82, 84, 85, 86, 88, 89, 90, 92, 93, 94, 96,
   97, 99,100,102,103,105,106,108,109,111,112,114,115,117,119,120,
  122,124,125,127,129,130,132,134,136,137,139,141,143,145,146,148,
  150,152,154,156,158,160,162,164,166,168,170,172,174,176,178,180,
  182,184,186,188,191,193,195,197,199,202,204,206,209,211,213,215,
  218,220,223,225,227,230,232,235,237,240,242,245,247,250,252,255};

// Fun routines
uint32_t Wheel(uint8_t WheelPos);
void rainbow(uint8_t wait);
void rainbowCycle(uint8_t wait);
void rainbowStep(uint8_t step);
void colorWipe(bool dir, uint32_t c, uint8_t wait);
uint32_t dimColor(uint32_t color, uint8_t width);
uint32_t colorWheel(uint8_t WheelPos);
void knightRider(uint16_t cycles, uint16_t speed, uint8_t width, uint16_t first, uint16_t last, uint32_t color);
void theaterChaseRainbow(int wait);

void ledsInitHW(uint16_t n, uint8_t o);
void ledsInitSW(uint16_t n, uint32_t dp, uint16_t d, uint32_t cp, uint16_t c, uint8_t o);
void updateLength(uint16_t n);
void ledsEnd(void);                                  // Destructor
void ledsBegin(void);                                // Prime pins/SPI for output
void ledsClear(void);                                // Set all pixel data to zero
void ledsSetBrightness(uint8_t);                     // Set global brightness 0-255
void ledsSetPixelColor(uint16_t n, uint32_t c);
void ledsSetPixelColorRGB(uint16_t n, uint8_t r, uint8_t g, uint8_t b);
void ledsUpdate(void);                               // Issue color data to strip

uint32_t ledsColor(uint8_t r, uint8_t g, uint8_t b); // R,G,B to 32-bit color
uint32_t ledsGetPixelColor(uint16_t n);              // Return 32-bit pixel color

uint16_t ledsNumPixels(void);                        // Return number of pixels

uint8_t ledsGetBrightness(void);                     // Return global brightness
uint8_t *ledsGetPixels(void);                        // Return pixel data pointer

/**
 * A gamma-correction function for 32-bit packed RGB or WRGB
 * colors. Makes color transitions appear more perceptially correct.
 */
uint32_t ledsColorHSV2(uint16_t hue, uint8_t sat, uint8_t val);
uint32_t ledsGamma32(uint32_t x);

#endif // _ADAFRUIT_DOT_STAR_H_
